Add further validation to prometheus receivers' validate method

rashmichandrashekar opened the pull request:
This PR extends the prometheus receivers' config validation method to add a few more checks similar to the promtool that are relevant to the receiver (https://github.com/open-telemetry/opentelemetry-collector/blob/d3000232f8f63566a4acb625943e0731db875b42/receiver/prometheusreceiver/config.go#L55) . The checks added are-
1. Check if Authorization credentials file exists.
2. Check TLS config - Certfile and key file specified actually exists. And one doesn't exist without the other.
3. Does the TLS config validation(as in 2 above) under Kubernetes SD config
4. For File SD config, unmarshals the files in the array and makes sure that they are in supported format( json or yaml ) and extracts target groups out of them.

Tracking issue - https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/4862

**Testing:** Have tested this out on a kubernetes cluster to make sure all the validation checks are run and relevant errors are returned when the scrape config is invalid.
I have included unit tests in this PR to test each of the checks added to make sure the expected errors are returned.

rashmichandrashekar committed:
Adding more validation

rashmichandrashekar committed:
auth test

rashmichandrashekar committed:
adding more tests

rashmichandrashekar committed:
kubernetes sd test

rashmichandrashekar committed:
adding more tests

rashmichandrashekar committed:
Merge pull request #1 from rashmichandrashekar/rashmi/promreceiver-config

rashmichandrashekar commented:
@dashpole - Could you please take a look at this PR? This is for the issue - https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/4862

codeboten commented on receiver/prometheusreceiver/config_test.go:
```diff
@@ -161,3 +161,147 @@ func TestRejectUnsupportedPrometheusFeatures(t *testing.T) {
+	require.Equal(t, wantErrMsg, gotErrMsg)
+}
+
+func TestTLSConfigCertFileExists(t *testing.T) {
```
Nit: This is testing what happens when the cert file is non-existent, the title of the test makes me think it's testing the opposite.

codeboten commented on receiver/prometheusreceiver/config_test.go:
```diff
@@ -161,3 +161,147 @@ func TestRejectUnsupportedPrometheusFeatures(t *testing.T) {
+	require.Equal(t, wantErrMsg, gotErrMsg)
+}
+
+func TestTLSConfigKeyFileExists(t *testing.T) {
```
Same nit as previous comment

codeboten commented on receiver/prometheusreceiver/config_test.go:
```diff
@@ -161,3 +161,147 @@ func TestRejectUnsupportedPrometheusFeatures(t *testing.T) {
+	require.Equal(t, wantErrMsg, gotErrMsg)
+}
+
+func TestTLSConfigCertFileWithNoKeyFile(t *testing.T) {
```
```suggestion
func TestTLSConfigCertFileWithoutKeyFile(t *testing.T) {
```

codeboten commented on receiver/prometheusreceiver/testdata/dummy-sd-config.json:
```diff
@@ -0,0 +1,12 @@
+[
```
Please rename this file to indicate it contains a nil target group.

codeboten reviewed (commented):
Great contribution, only a few minor comments.

odeke-em commented on receiver/prometheusreceiver/config.go:
```diff
@@ -50,6 +58,66 @@ type Config struct {
+	if fn == "" {
+		return nil
+	}
+	_, err := os.Stat(fn)
```
Yup, it can return permission errors too.

rashmichandrashekar committed:
addressing PR comments

rashmichandrashekar committed:
Merge pull request #2 from rashmichandrashekar/rashmi/pr-comments

rashmichandrashekar committed:
Merge pull request #3 from rashmichandrashekar/rashmi/promreceiver-config

Adressing PR comments - https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/5052

rashmichandrashekar commented:
> Great contribution, only a few minor comments.

@codeboten - Thanks for taking a look. I have addressed all your comments. please feel free to resolve them or let me know if the fixes don't address the comments and I can fix them.

rashmichandrashekar commented on receiver/prometheusreceiver/config.go:
```diff
@@ -50,6 +58,66 @@ type Config struct {
+	return nil
+}
+
+func checkSDFile(filename string) error {
```
Good point. I looked into the file.go file for the code reference, the method doesn't look like something that can be exported. Have added a comment like you suggested. Please take a look.

dashpole commented on receiver/prometheusreceiver/config.go:
```diff
@@ -93,6 +163,42 @@ func (cfg *Config) Validate() error {
+						}
+						continue
+					}
+					fmt.Printf("WARNING: file %q for file_sd in scrape job %q does not exist\n", file, sc.JobName)
```
Should we be returning an error here instead of printing a warning?  If we still want to print, should we be using a logger?

rashmichandrashekar replied:
Makes sense. I have updated it to return error instead of logging to keep it consistent. Have fixed another test which had a non existent file here -https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/5052/files#diff-91a13efe7ac836919f7e9db1752692243725fbf17683c645dedcb27a6cfff4ac  Please take a look.

rashmichandrashekar committed:
fixing PR comment related to returning error when file doesnt exist in sd config

rashmichandrashekar committed:
Merge pull request #4 from rashmichandrashekar/rashmi/pr-comments

rashmichandrashekar committed:
Merge pull request #5 from rashmichandrashekar/rashmi/pr-comments

rashmichandrashekar commented:
> > Great contribution, only a few minor comments.
> 
> @codeboten - Thanks for taking a look. I have addressed all your comments. please feel free to resolve them or let me know if the fixes don't address the comments and I can fix them.



> > Great contribution, only a few minor comments.
> 
> @codeboten - Thanks for taking a look. I have addressed all your comments. please feel free to resolve them or let me know if the fixes don't address the comments and I can fix them.

Also, I am trying to get this merged for the next release so that we can pick up the latest collector version. Is it possible to take a look and let me know if there is anything else i need to fix, so that i can get this merged sooner?

codeboten reviewed (approved):
Thanks for addressing my comments 👍

rashmichandrashekar committed:
Fixing lint errors

rashmichandrashekar committed:
Merge pull request #6 from rashmichandrashekar/rashmi/pr-comments

rashmichandrashekar committed:
Merge pull request #7 from rashmichandrashekar/rashmi/promreceiver-config

Fixing lint errors

rashmichandrashekar commented:
@dashpole @codeboten @Aneurysm9 - Thanks for approving the PR! I see that once of the checks build-publish is failing, unrelated to my changes, and i also see this for a few other PRs. Could you pls let me know how I can work around this to get this merged?
